src/sage/rings/function_field/jacobian_base.py

```diff
@@ -0,0 +1,797 @@
+# ****************************************************************************
+
+import math
+
+from sage.arith.misc import integer_floor, integer_ceil
+
+from sage.structure.parent import Parent
+from sage.structure.element import ModuleElement
+
+from sage.categories.commutative_additive_groups import CommutativeAdditiveGroups
+from sage.categories.schemes import Schemes
+from sage.categories.pushout import ConstructionFunctor, pushout
+
+from sage.rings.integer_ring import IntegerRing
+from sage.rings.integer import Integer
+
+
+class JacobianPoint_base(ModuleElement):
+    """
+    Abstract base class of points of Jacobian groups.
+    """
+    pass
+
+
+class JacobianPoint_finite_field_base(JacobianPoint_base):
+    """
+    Points of Jacobians over finite fields.
+    """
+    def order(self):
+        """
+        Return the order of this point.
+
+        EXAMPLES::
+
+            sage: P2.<x,y,z> = ProjectiveSpace(GF(29), 2)
+            sage: C = Curve(x^3 + 5*z^3 - y^2*z, P2)
+            sage: F = C.function_field()
+            sage: h = C.function(y/x).divisor_of_poles()
+            sage: J = C.jacobian(model='km_large', base_div=h)
+            sage: G = J.group()
+            sage: b = F.get_place(1)
+            sage: pl = C([-1,2,1]).place()
+            sage: p = G.point(pl - b)
+            sage: p.order()
+            15
+
+        ALGORITHM: Shanks' Baby Step Giant Step
+        """
+        G = self.parent()
+        B = G._bound_on_order()
+        q = integer_ceil(B.sqrt())
+        zero = G.zero()
+
+        # baby steps
+        b = [zero]
+        g = self
+        for i in range(q - 1):
+            if g == zero:
+                return i + 1
+            b.append(g)
+            g = g + self
+
+        # giant steps
+        g0 = self.multiple(-q)
+        g = g0
+        for i in range(q - 1):
+            for r in range(q):
+                if g == b[r]:
+                    return q * (i + 1) + r
+            g = g + g0
+
+        # order is neither smaller or nor larger than this
+        return q**2
+
+    def frobenius(self):
+        """
+        Return the image of the point acted by the Frobenius automorphism.
+
+        EXAMPLES::
+
+            sage: k = GF(7)
+            sage: A.<x,y> = AffineSpace(k,2)
+            sage: C = Curve(y^2 + x^3 + 2*x + 1).projective_closure()
+            sage: J = C.jacobian(model='hess')
+            sage: G1 = J.group()
+            sage: G1.order()
+            11
+            sage: K = k.extension(3)
+            sage: G3 = J.group(K)
+            sage: pts1 = G1.get_points(11)
+            sage: pts3 = G3.get_points(12)
+            sage: pt = next(pt for pt in pts3 if pt not in pts1)
+            sage: pt.frobenius() == pt
+            False
+            sage: pt.frobenius().frobenius().frobenius() == pt
+            True
+        """
+        G = self.parent()
+        return G._frobenius_on(self)
+
+
+class Jacobians(Schemes):
```

Comment:
Is this done correctly? There should be tests added that illustrate this.

It seems that commonly subcategories are defined in a different way, by giving them a method `super_categories`.

Reply:
Done.